### PR TITLE
Workaround for ignoring some start parameters

### PR DESCRIPTION
**Pull request by 6hundreds**
Fixes https://github.com/gradle/gradle/issues/26360

In fact, we can't use `InstrumentedInputAccessListener.systemPropertyQueried` even if transform `cid` to system property. 
`cid` is has passing as argument to daemon so finally it getting into `startParameters`. So even if we change `cid` to sysProp, it still will be presented in start parameters and will invalidate the cache. 

This PR is introducing workaround for ignoring start parameters. This allowing to filter `cid` from start parameters and have a CC hit

### CODE REVIEW

**Comment by 6hundreds** [comment]
@bamboo I also wondered how `cid` will be matched in this case, if after CC hit daemon and client will use different `cid`s

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/initialization/ConfigurationCacheStartParameter.kt [hunk 1]
```diff
@@ -59,6 +60,7 @@ class ConfigurationCacheStartParameter(
 
     val gradleProperties: Map<String, Any?>
         get() = startParameter.projectProperties
+            .filter { !Workarounds.isIgnoredStartParameterProperty(it.key) }
```
Comment (bamboo) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/initialization/ConfigurationCacheStartParameter.kt:63]: ❓ Why? System properties are not part of `projectProperties` are they?

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheAllowedStartParametersIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+class ConfigurationCacheAllowedStartParametersIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    def "changing of kotlin dsl cid property doesn't invalidates cache entry"() {
+        given:
+        def kotlinDslCidProp = "org.gradle.kotlin.dsl.provider.cid"
+        def configurationCache = newConfigurationCacheFixture()
+
+        when:
+        configurationCacheRun "tasks", "-D$kotlinDslCidProp = 24"
```
Comment (bamboo) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheAllowedStartParametersIntegrationTest.groovy:27]: ❓ I guess in this case you want to test passing project properties, right? 
```suggestion
        configurationCacheRun "tasks", "-P$kotlinDslCidProp = 24"
```

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheAllowedStartParametersIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+        then:
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheRun "tasks", "-D$kotlinDslCidProp = 42"
```
Comment (bamboo) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheAllowedStartParametersIntegrationTest.groovy:33]:
```suggestion
        configurationCacheRun "tasks", "-P$kotlinDslCidProp = 42"
```

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/initialization/ConfigurationCacheStartParameter.kt (hunk 1 above)
Comment (bamboo) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/initialization/ConfigurationCacheStartParameter.kt:63]: I see now that this PR is about preserving the cid as a project property.

**Comment by 6hundreds** [comment]
@bamboo ship it? 😄

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheAllowedStartParametersIntegrationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+class ConfigurationCacheAllowedStartParametersIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    def "changing of kotlin dsl cid property doesn't invalidates cache entry"() {
+        given:
+        def kotlinDslCidProp = "org.gradle.kotlin.dsl.provider.cid"
+        def configurationCache = newConfigurationCacheFixture()
+
+        when:
+        configurationCacheRun "tasks", "-P$kotlinDslCidProp= 24"
```
Comment (h0tk3y) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheAllowedStartParametersIntegrationTest.groovy:27]: 💭 Is the whitespace in `-P$kotlinDslCidProp= 24` parsed properly? 
I thought: could it be parsed as an empty value + whitespace + dangling argument `24`?

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/initialization/ConfigurationCacheStartParameter.kt [hunk 5]
```diff
@@ -59,6 +60,7 @@ class ConfigurationCacheStartParameter(
 
     val gradleProperties: Map<String, Any?>
         get() = startParameter.projectProperties
+            .filter { !Workarounds.isIgnoredStartParameterProperty(it.key) }
```
Comment (h0tk3y) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/initialization/ConfigurationCacheStartParameter.kt:63]:
```suggestion
            .filterKeys { !Workarounds.isIgnoredStartParameterProperty(it) }
```

**Review by h0tk3y — APPROVED** [review]

**Comment by 6hundreds** [comment]
@bot-gradle test and merge

**Review by bamboo — APPROVED** [review]
💯

**Pull request merged**